File: 79-word-search/79-word-search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        m, n = len(board), len(board[0])
        length = len(word)
        visited = dict()
        record = dict()
        def find(i,j, k):
            if (i,j,k) in record:
                return record[(i,j,k)]
            dire = [(0,1),(0,-1),(1,0),(-1,0)]
            if k == length:
                return True
            if word[k] != board[i][j]:
                return False
            for nxt in dire:
                newX, newY = i+nxt[0], j + nxt[1]
                if 0 <= newX < m and 0 <= newY < n and (newX, newY) not in visited:
                    visited[(newX,newY)] = 1
                    if find(newX, newY, k+1):
                        record[(newX,newY,k+1)] = True
                        record[(i,j,k)] = True
                        return record[(newX,newY,k+1)]
                    del visited[(newX,newY)]
            if k+1 == length:
                record[(i,j,k)] = True
                return True
      #      else:
            
      #          record[(i,j,k)] = False        
            return False
        for i in range(m):
            for j in range(n):
                visited[(i,j)] = 1
                if find(i,j, 0):
                    return True
                del visited[(i,j)]
        return False
                        
                    
            
        

File: 79-word-search/79-word-search.java
class Solution {
    int[][] directions = new int[][]{{1,0}, {-1,0},{0,1},{0,-1}};
    int nrow;
    int ncol;
    int wordLen;
    Set<Pair> visited = new HashSet<>();
    public boolean exist(char[][] board, String word) {
        nrow = board.length;
        ncol = board[0].length;
        wordLen = word.length();
        for(int i = 0; i < nrow; i++)
            for(int j = 0; j < ncol; j++){
                char curChar = board[i][j];
                Pair key = new Pair(i, j);
                if(curChar == word.charAt(0)){
                    visited.add(key);
                    if (dfs(board, i, j, word, 1)) return true;
                    visited.remove(key);
                } 
            }
        return false;
    }
    public boolean dfs(char[][] board, int row, int col, String word, int wordIdx){
        if(wordIdx == wordLen)return true;
        for(int[] direction : directions){
            int x = row + direction[0], y = col + direction[1];
            Pair key = new Pair(x, y);
            if( x < nrow && x >= 0 && y >= 0 && y < ncol && !visited.contains(key) && board[x][y] == word.charAt(wordIdx)){
                visited.add(key);
                if (dfs(board, x, y, word, wordIdx + 1))return true;
                visited.remove(key);
                
            }
        }
        return false;
    }
}